src/main/scala/mesosphere/marathon/core/storage/PersistenceStore.scala

```diff
@@ -0,0 +1,254 @@
+      case None =>
+        Failure(new StoreCommandFailedException(s"Unable to update $id as it doesn't exist"))
+    }).map(_.get)(CallerThreadExecutionContext.callerThreadExecutionContext)
+  }
+
+  /**
+    * Delete the value at the given Id, idempotent
+    *
+    * @return A future indicating whether the value was deleted (or simply didn't exist). Underlying storage issues
+    *         will fail the future with [[StoreCommandFailedException]]
+    */
+  def delete[Id, V](k: Id)(implicit ir: IdResolver[Id, K, V, Serialized]): Future[Done] = {
+    rawDelete(ir.toStorageId(k))
+  }
+
+  /**
+    * @return A source of _all_ keys in the Persistence Store (which can be used by a
+    *         [[mesosphere.marathon.core.storage.impl.CachingPersistenceStore]] to populate the
+    *         cache completely on startup.
+    */
+  protected[storage] def keys(): Source[K, NotUsed]
+
+  /**
+    * Delete the value at the given id.
+    * @return A future that completes when the value was removed or didn't exist.
+    *         The future may fail with a [[StoreCommandFailedException]] if there is an underlying storage issue.
+    */
+  protected def rawDelete(id: K): Future[Done]
+
+  /**
+    * Get a list of all immediate children under the given parent.
+    */
+  protected def rawIds(parent: K): Source[K, NotUsed]
+
+  /**
+    * Get the data, if any, for the given primary id and value type.
+    *
+    * @return A future representing the data at the given Id, if any exists.
+    *         If there is an underlying storage problem, the future should fail with [[StoreCommandFailedException]]
+    */
+  protected[storage] def rawGet(id: K): Future[Option[Serialized]]
+
+  /**
+    * Set the value at the given id.
+    *
+    * @return A future that completes when the value has been updated to the new value
+    *         or fail with [[StoreCommandFailedException]] if there is an underlying storage issue or the value
+    *         didn't exist.
+    */
+  protected def rawSet(id: K, v: Serialized): Future[Done]
+
+  /**
+    * Create the value at the given id.
+    *
+    * @return A future that completes when the value has been associated with the Id
+    *         or fail with [[StoreCommandFailedException]] if there is an underlying storage issue or the
+    *         value already existed.
+    */
+  protected def rawCreate(id: K, v: Serialized): Future[Done]
+
+  private def createOrUpdate[Id, V](
+    id: Id,
+    change: Option[V] => Try[V])(implicit
+    ir: IdResolver[Id, K, V, Serialized],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V],
+    ctx: ExecutionContext = ExecutionContext.global): Future[Option[V]] = async {
+    val path = ir.toStorageId(id)
+    val old = await(get(id))
+    change(old) match {
+      case Success(newValue) =>
+        val serialized = await(Marshal(newValue).to[Serialized])
+        old match {
+          case Some(_) =>
+            await(rawSet(path, serialized))
+            old
```

Comment:
Not 100% sure about the expected return value, but I would have guessed it should be `newValue` not `old`.